### PR TITLE
Show pending/invalid runs better

### PR DESCRIPTION
**Pull request by Hexcles**
Use `<vaadin-grid>` instead of `<table>` to allow us show details, select
columns, etc.

## Review Information

Try the staging instance.

https://better-status-dot-wptdashboard-staging.appspot.com/status

### CODE REVIEW

**Comment by Hexcles** [comment]
TIL more about flexbox and shadow DOM styling.

File: webapp/components/wpt-processor.js [hunk 1]
```diff
@@ -38,39 +43,64 @@ class WPTProcessor extends LoadingState(PolymerElement) {
...
+
+      <vaadin-grid aria-label="Test runs" items="[[testRuns]]">
+        <vaadin-grid-column auto-width header="ID">
+          <template>[[item.id]]</template>
+        </vaadin-grid-column>
+        <!-- TODO(Hexcles): Show this column by default when we have data. -->
+        <vaadin-grid-column auto-width header="GitHub Check Run" hidden>
+          <template>[[item.check_run_id]]</template>
+        </vaadin-grid-column>
+        <vaadin-grid-column auto-width header="SHA">
```
[the remark this answers is not in the document]
Comment (Hexcles) [reply webapp/components/wpt-processor.js]: That's a bug and I have a fix coming :)

File: webapp/components/wpt-processor.js [hunk 2]
```diff
@@ -38,39 +43,64 @@ class WPTProcessor extends LoadingState(PolymerElement) {
...
-            <td title="[[run.error]]">[[ run.stage ]]</td>
-            <td>[[ run.uploader ]]</td>
-          </tr>
+    <template is="dom-if" if="[[testRuns.length]]" on-dom-change="refreshContextMenu">
+      <x-data-provider data-provider="[[testRuns]]"></x-data-provider>
+
+      <div id="note">
+        <p>Note: timestamps are displayed in your local timezone.</p>
+        <vaadin-context-menu open-on="click">
+          <vaadin-button theme="icon" aria-label="Hide column menu">
```
[the remark this answers is not in the document]
Comment (Hexcles) [reply webapp/components/wpt-processor.js]: Yes. And sorry what's the question?

File: webapp/components/wpt-processor.js [hunk 3]
```diff
@@ -38,39 +43,64 @@ class WPTProcessor extends LoadingState(PolymerElement) {
...
+          <template>[[timestamp(item.created)]]</template>
+        </vaadin-grid-column>
+        <vaadin-grid-column auto-width header="Uploaded">
+          <template>[[timestamp(item.updated)]]</template>
+        </vaadin-grid-column>
+        <vaadin-grid-column auto-width header="Stage">
+          <template>[[item.stage]]</template>
+        </vaadin-grid-column>
+
+        <vaadin-grid-column autowidth header="Show error">
```
[the remark this answers is not in the document]
Comment (Hexcles) [reply webapp/components/wpt-processor.js]: We could, but all invalid runs *should* have errors.

File: webapp/components/wpt-processor.js [hunk 4]
```diff
@@ -38,39 +43,64 @@ class WPTProcessor extends LoadingState(PolymerElement) {
...
+          <template class="header">
+            Show error <vaadin-checkbox aria-label="Show all" on-checked-changed="toggleAllDetails" id="show-all"></vaadin-checkbox>
+          </template>
+          <template>
+            <vaadin-checkbox class="show-details" aria-label$="Show error for [[item.id]]" checked="{{detailsOpened}}"></vaadin-checkbox>
```
[the remark this answers is not in the document]
Comment (Hexcles) [reply webapp/components/wpt-processor.js]: Ditto.

File: webapp/components/wpt-processor.js [hunk 5]
```diff
@@ -38,39 +43,64 @@ class WPTProcessor extends LoadingState(PolymerElement) {
...
-            <td class="time-ago">[[ timeAgo(run.updated) ]]</td>
-            <td class="timestamp">[[ timestamp(run.created) ]]</td>
-            <td class="time-ago">[[ timeAgo(run.created) ]]</td>
-            <td title="[[run.error]]">[[ run.stage ]]</td>
-            <td>[[ run.uploader ]]</td>
-          </tr>
+    <template is="dom-if" if="[[testRuns.length]]" on-dom-change="refreshContextMenu">
+      <x-data-provider data-provider="[[testRuns]]"></x-data-provider>
+
+      <div id="note">
```
[the remark this answers is not in the document]
Comment (Hexcles) [reply webapp/components/wpt-processor.js]:
```suggestion
      <div id="before-grid">
```

File: webapp/components/wpt-processor.js [hunk 6]
```diff
@@ -4,24 +4,29 @@
...
-      table {
-        width: 100%;
-        max-width: 1200px;
+      :host {
+        display: flex;
+        flex-direction: column;
       }
-      td {
-        text-align: center;
+      #note p {
```
Comment (Hexcles) [inline webapp/components/wpt-processor.js]:
```suggestion
      #before-grid p {
```

File: webapp/components/wpt-processor.js [hunk 7]
```diff
@@ -4,24 +4,29 @@
...
+        float: left;
+      }
+      #note vaadin-context-menu {
```
Comment (Hexcles) [inline webapp/components/wpt-processor.js]:
```suggestion
      #before-grid vaadin-context-menu {
```

File: webapp/components/wpt-processor.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (stephenmcgruer) [reply webapp/components/wpt-processor.js]: Ohhhh, is 'Hide column' here a single concept? I was reading this as {verb, noun}, i.e. that clicking on this button would hide (verb) the column menu (noun), and was concerned that this button only does that when the menu is open, and does the opposite when it is closed.

File: webapp/components/wpt-processor.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (stephenmcgruer) [reply webapp/components/wpt-processor.js]: What about pending runs? I thought this was the file for both pending + invalid, maybe I got confused.

File: webapp/components/wpt-processor.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (Hexcles) [reply webapp/components/wpt-processor.js]: Ahh the ARIA labels were from the sample snippet and it didn't occur to me that there was ambiguity. Thanks!

```suggestion
          <vaadin-button theme="icon" aria-label="Select columns">
```

File: webapp/components/wpt-processor.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (Hexcles) [reply webapp/components/wpt-processor.js]: Yes, it was (and still is) intended for both, and I apparently forgot about that.

I'm going to claim pending runs are an edge case (we usually only have fewer than a handful of them at any given moment) and leave the UI as is ;)

[the remark this answers is not in the document]
Comment (stephenmcgruer) [reply webapp/components/wpt-processor.js]: SGTM.

**Pull request merged**